### ISSUE TITLE
OMEMO error with my other client

### ISSUE THREAD

**Issue by aereaux** [profanity-im/profanity#1436]
## Expected Behavior
When I send an OMEMO-encrypted message to a group chat or to a contact directly, they should be able to see the message (and my other devices should as well).  I am using profanity on one device and conversations on another, and my other contact is using just conversations.

## Current Behavior
When I send an OMEMO-encrypted message to a group chat that includes a certain one of my contacts, or to that person directly, they only get a "message not encrypted for this device" message.  I also get the same message on my other device when using a group chat, but it works (on my device) when sending a message to that contact directly.

## Possible Solution
No idea.

## Steps to Reproduce (for bugs)
I have no idea how to reproduce this, other than that it is very reproducible with my contact and me.  I can provide more debugging information if needed.

1. Profanity opens the group chat automatically, and it defaults to omemo
2. I send a message in the group chat's tab

## Context
I am still able to talk encrypted with my contact through my conversations instance, but it is a little annoying and worrying.

I was also able to do some debugging with the help of the profanity MUC.  I found that this message gets output to the profanity log:
```
prof: WRN: OMEMO: cannot find device ids for <contact jid>
```
I also found a weird thing when I enabled debug logging.  There are two other people in this group chat, not including me.  I see the stanza sent in the debug log, and there are four `key` xml tags.  Each of the devices of the contact that works is listed twice, with none for my other device or the one device of the contact that gets the error message.  Not sure if this is intended, or if this could be (part of) the problem.

## Environment
* Give us the version and build information output generated by `profanity -v`
```
Profanity, version 0.9.5
Copyright (C) 2012 - 2019 James Booth <boothj5web@gmail.com>.
Copyright (C) 2019 - 2020 Michael Vetter <jubalh@iodoru.org>.
License GPLv3+: GNU GPL version 3 or later <https://www.gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Build information:
XMPP library: libmesode
Desktop notification support: Enabled
OTR support: Enabled (libotr 4.1.1)
PGP support: Enabled (libgpgme 1.14.0)
OMEMO support: Enabled
C plugins: Enabled
Python plugins: Enabled (3.8.6)
GTK icons: Disabled
````
* If you could not yet build profanity, mention the revision you try to build from
* Operating System/Distribution: Arch Linux
* glib version: glib2 2.66.2-1

**Comment by jubalh** [comment]
OMEMO Gang: @paulfariello @wstrm @DebXWoody :-)

**Comment by paulfariello** [comment]
Did you trust your contact keys on profanity with `/omemo trust ` command?

**Comment by aereaux** [comment]
Yep, sorry for not mentioning that.  Both of my contacts keys are listed as trusted when I do `/omemo fingerprint <contact>`.  I'm not sure why there are two keys.  Possibly they used another client in the past, I don't remember.

**Comment by aereaux** [comment]
Sorry for not figuring this out sooner, but it looks like this is (partially) fixed in master.  Profanity now chooses the first device list if there is not one marked "current".  I still get the error about not encrypted for this device on my conversations client.  Should Profanity encrypt for my other devices, or is that expected?

**Comment by jubalh** [comment]
>  I still get the error about not encrypted for this device on my conversations client. Should Profanity encrypt for my other devices, or is that expected?

omemo gang ping: @paulfariello @DebXWoody @wstrm

**Comment by jubalh** [comment]
@aereaux as far as I know omemo code in the git version changed. Would you mind testing whether it works with latest git?

**Comment by aereaux** [comment]
I've been using the git version for a while after I filed this issue, and it appears not to have fixed it.  Currently testing another PR, but after that I can use the latest latest git to verify.

**Comment by jubalh** [comment]
@aereaux maybe the debugging section of https://github.com/profanity-im/blog/blob/master/content/post/omemo.md may help?

**Comment by aereaux** [comment]
A bit of debugging info: my device is listed in trust.txt and in the devicelist response, but the rid is not included when I send a message.

**Comment by ElDifinitivo** [comment]
I'm having the same issue, but with 1:1 chats (haven't tried MUC).

I use both profanity and Conversations, and my contact uses Convesations and Dino on desktop. I've made sure to trust both my contacts' fingerprints and even both of mine with `/omemo trust <fingerprint>`. Conversations also has enabled my profanity fingerprint. My contact properly decrypts messages, profanity properly shows the message carbons when sent from Conversations, but anything sent from profanity shows "Message was not encrypted for this device" on my Conversations client.

I've read through 2 similar issues in the Github for Conversations and it all points to profanity not properly encrypting for the other client.

The debug logs for the sent message also contain a `<body>` that states "You received a message encrypted with OMEMO but your client doesn't support OMEMO.

```
01/02/2021 09:50:46: conn: DBG: SENT: <message id="$MESSAGE_ID" to="CONTACT@DOMAIN" type="chat"><encrypted xmlns="eu.siacs.conversations.axolotl"><header sid="$HEADER"><key rid="$RID">$RID_KEY</key><key rid="$RID_2">$RID_2_KEY</key><iv>$IV</iv></header><payload>Vt0l</payload></encrypted><body>You received a message encrypted with OMEMO but your client doesn't support OMEMO.</body><active xmlns="http://jabber.org/protocol/chatstates"/><store xmlns="urn:xmpp:hints"/></message>
```

**Comment by jubalh** [comment]
@wstrm any idea?

**Comment by jpds** [comment]
I just started using Profanity and I knew from experience that I needed to trust my own OMEMO fingerprints for message carbons to my own/contacts' devices to work.

Issues I had from a Profanity perspective:

* `/help omemo` - tells you how to generate a OMEMO key, doesn't tell you that `/omemo fingerprint` can be used to view your own fingerprints (just contacts)
* When you are in a chat `/omemo fingerprint` only shows the contact fingerprints, not the ones you additionally trust for yourself
* `/omemo fingerprint` in the console tells you
```
You can trust it with '/omemo trust <fingerprint>'
```
...but then when you try and do this for your own fingerprint, you get:
```
You must be in a regular chat window to trust a device without providing the contact.
```
* Rather than being told that you have to do `/omemo trust your.own.jid@domain.com fingerprint` in the console to trust your own keys.

**Comment by kaffeekanne** [comment]
@jpds as a workarount you can start a chat with yourself.
```
/omemo start <yourjid>
/omemo fingerprint
/omemo trust <fingerprint>
...
```
It is hacky, but you have to do it only once. Don't trust the fingerprint of your profanity instance, as you should not encrypt for the sending resource. It's somewhere in the XEP.

**Comment by mwuttke97** [comment]
I'm experiencing the same issue. It's quite annoying, since I can't know whether my recipent can actually decrypt the message, unless I look into the log, for entries like this:

> DBG: Input received: text
> prof: WRN: [OMEMO] cannot find device ids for `you@B.host`
> prof: ERR: [OMEMO] cannot encrypt key for `me@A.host` device id myid1
> prof: ERR: [OMEMO] cannot encrypt key for `me@A.host` device id myid2

I noticed that uppon creating a new chat window with `B@B.host`, two 'get-device-list' requests were send to `you@B.host` and `me@A.host`. I highly suspect that there's a race condition going on. This would explain why it's sometimes working and sometimes not. (If it's not working, the chat has to be restarted again without bad luck.)
